receiver/wavefrontreceiver/receiver_test.go

```diff
@@ -0,0 +1,170 @@
+	require.NoError(t, err)
+
+	mh := component.NewMockHost()
+	require.NoError(t, rcvr.Start(mh))
+	defer rcvr.Shutdown()
+
+	tests := []struct {
+		name string
+		msg  string
+		want []*metricspb.Metric
+	}{
+		{
+			name: "single.line",
+			msg:  "single.metric 1 1582231120 source=e2e\n",
+			want: []*metricspb.Metric{
+				buildMetric(
+					metricspb.MetricDescriptor_GAUGE_INT64,
+					"single.metric",
+					[]string{"source"},
+					[]string{"e2e"},
+					&metricspb.Point{
+						Timestamp: &timestamp.Timestamp{Seconds: 1582231120},
+						Value:     &metricspb.Point_Int64Value{Int64Value: 1},
+					},
+				),
+			},
+		},
+		{
+			name: "single.line.no.newline",
+			msg:  "single.metric 1 1582231120 source=e2e",
+			want: []*metricspb.Metric{
+				buildMetric(
+					metricspb.MetricDescriptor_GAUGE_INT64,
+					"single.metric",
+					[]string{"source"},
+					[]string{"e2e"},
+					&metricspb.Point{
+						Timestamp: &timestamp.Timestamp{Seconds: 1582231120},
+						Value:     &metricspb.Point_Int64Value{Int64Value: 1},
+					},
+				),
+			},
+		},
+		{
+			name: "multiple.lines",
+			msg:  "m0 0 1582231120 source=s0\nm1 1 1582231121 source=s1\nm2 2 1582231122 source=s2\n",
+			want: []*metricspb.Metric{
+				buildMetric(
+					metricspb.MetricDescriptor_GAUGE_INT64,
+					"m0",
+					[]string{"source"},
+					[]string{"s0"},
+					&metricspb.Point{
+						Timestamp: &timestamp.Timestamp{Seconds: 1582231120},
+						Value:     &metricspb.Point_Int64Value{Int64Value: 0},
+					},
+				),
+				buildMetric(
+					metricspb.MetricDescriptor_GAUGE_INT64,
+					"m1",
+					[]string{"source"},
+					[]string{"s1"},
+					&metricspb.Point{
+						Timestamp: &timestamp.Timestamp{Seconds: 1582231121},
+						Value:     &metricspb.Point_Int64Value{Int64Value: 1},
+					},
+				),
+				buildMetric(
+					metricspb.MetricDescriptor_GAUGE_INT64,
+					"m2",
+					[]string{"source"},
+					[]string{"s2"},
+					&metricspb.Point{
+						Timestamp: &timestamp.Timestamp{Seconds: 1582231122},
+						Value:     &metricspb.Point_Int64Value{Int64Value: 2},
+					},
+				),
+			},
+		},
+	}
+	for _, tt := range tests {
+		conn, err := net.Dial("tcp", addr)
+		require.NoError(t, err)
+
+		numMetrics := strings.Count(tt.msg, "\n")
+		if numMetrics == 0 {
+			numMetrics = 1
+		}
+		waitableConsumer.Add(numMetrics)
+		n, err := fmt.Fprint(conn, tt.msg)
+		assert.Equal(t, len(tt.msg), n)
+		assert.NoError(t, err)
+
+		require.NoError(t, conn.Close())
+		waitableConsumer.Wait()
+
+		got := waitableConsumer.PullReceivedMetrics()
+		assert.Equal(t, tt.want, got)
+	}
+}
+
+type waitableMetricsConsumer struct {
+	sync.WaitGroup
+	mtx     sync.Mutex
+	metrics []*metricspb.Metric
+}
+
+var _ (consumer.MetricsConsumer) = (*waitableMetricsConsumer)(nil)
+
+func (w *waitableMetricsConsumer) ConsumeMetricsData(ctx context.Context, md consumerdata.MetricsData) error {
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+	defer w.Done()
```

Comment:
May be in this case defer is not that useful as the function is pretty simple? Till 1.14 defer still incurs a noticeable cost especially for hot paths.

Reply:
Fair enough